packages/world/src/systems/TerrainSystem.sol

```diff
@@ -0,0 +1,102 @@
+    require(ExploredChunk.get(chunkCoord.x, chunkCoord.y, chunkCoord.z) == address(0), "Chunk already explored");
+    // TODO: verify merkle proof
+    SSTORE2.writeDeterministic(chunkData, Terrain._getChunkSalt(chunkCoord));
+    ExploredChunk.set(chunkCoord.x, chunkCoord.y, chunkCoord.z, _msgSender());
+  }
+}
+
+library Terrain {
+  using SSTORE2 for address;
+  bytes1 constant _VERSION = 0x00;
+
+  /// @notice Get the terrain block type of a voxel coordinate.
+  /// @dev Assumes to be called from a root system.
+  function _getBlockType(VoxelCoord memory coord) internal view returns (uint8) {
+    return getBlockType(coord, address(this));
+  }
+
+  /// @notice Get the terrain block type of a voxel coordinate.
+  /// @dev Can be called from either a root or non-root system, but consumes slightly more gas.
+  function getBlockType(VoxelCoord memory coord) internal view returns (uint8) {
+    return getBlockType(coord, WorldContextConsumerLib._world());
+  }
+
+  /// @notice Get the terrain block type of a voxel coordinate.
+  function getBlockType(VoxelCoord memory coord, address world) internal view returns (uint8) {
+    ChunkCoord memory chunkCoord = _getChunkCoord(coord);
+    require(_isChunkExplored(chunkCoord, world), "Chunk not explored yet");
+
+    address chunkPointer = _getChunkPointer(chunkCoord, world);
+    bytes1 version = chunkPointer.readBytes1(0);
+    require(version == _VERSION, "Unsupported chunk encoding version");
+
+    uint256 index = _getBlockIndex(coord);
+    bytes1 blockType = chunkPointer.readBytes1(index);
+
+    return uint8(blockType);
+  }
+
+  /// @dev Get the chunk coordinate of a voxel coordinate
+  function _getChunkCoord(VoxelCoord memory chunkCoord) internal pure returns (ChunkCoord memory) {
+    return
+      ChunkCoord({
+        x: floorDiv(chunkCoord.x, CHUNK_SIZE),
+        y: floorDiv(chunkCoord.y, CHUNK_SIZE),
+        z: floorDiv(chunkCoord.z, CHUNK_SIZE)
+      });
+  }
+
+  /// @dev Get the relative coordinate of a voxel coordinate within a chunk
+  function _getRelativeCoord(VoxelCoord memory coord) internal pure returns (VoxelCoord memory) {
+    return
+      VoxelCoord({
+        x: int32(uint32(mod(coord.x, CHUNK_SIZE))),
+        y: int32(uint32(mod(coord.y, CHUNK_SIZE))),
+        z: int32(uint32(mod(coord.z, CHUNK_SIZE)))
+      });
+  }
+
+  /// @dev Get the index of a voxel coordinate within the encoded chunk
+  function _getBlockIndex(VoxelCoord memory coord) internal pure returns (uint256) {
+    VoxelCoord memory relativeCoord = _getRelativeCoord(coord);
+    return
+      VERSION_PADDING +
+      uint256(
+        int256(relativeCoord.x) * CHUNK_SIZE ** 2 + int256(relativeCoord.y) * CHUNK_SIZE + int256(relativeCoord.z)
+      );
+  }
+
+  /// @dev Get the salt for a chunk coordinate
+  function _getChunkSalt(ChunkCoord memory coord) internal pure returns (bytes32) {
+    return bytes32(uint256((int256(coord.x) << 64) | (int256(coord.y) << 32) | int256(coord.z)));
```

Comment:
Got a bit confused what it was doing here. I assume the goal here is to encode the coord as a bytes32? Is this done in this bit shifting way for gas reasons vs:
```
keccak256(abi.encodePacked(coord.x, coord.y, coord.z))
```

Reply:
Yeah correct, we just need a unique bytes32 salt per chunk coordinate, so we don't need to hash and can save some gas. Also using bitshift instead of abi.encodePacked here to save some gas. There is a test case that shows the equivalence of this operation with a different approach using abi.encodePacked